### ISSUE TITLE
self_managed_node_group creates invalid launch template when using network_interfaces

### ISSUE THREAD

**Issue by mark-nimbix** [terraform-aws-modules/terraform-aws-eks#1880]
## Description

When using the EKS module to create a self managed node group if you define a network_interface in the self_managed_node_group sub module it will create an invalid template per issue https://github.com/hashicorp/terraform-provider-aws/issues/4570. It looks like there needs to be a conditional for the vpc_security_group_ids value setting it to null when a network_interface is provided. I was able to get past the error and add the EFA interface_type I wanted by adding the following.
```
git diff
diff --git a/modules/self-managed-node-group/main.tf b/modules/self-managed-node-group/main.tf
index 1b7ca66..52bcb81 100644
--- a/modules/self-managed-node-group/main.tf
+++ b/modules/self-managed-node-group/main.tf
@@ -57,7 +57,7 @@ resource "aws_launch_template" "this" {
   key_name      = var.key_name
   user_data     = module.user_data.user_data

-  vpc_security_group_ids = compact(concat([try(aws_security_group.this[0].id, "")], var.vpc_security_group_ids))
+  vpc_security_group_ids = length(var.network_interfaces) > 0 ? null :  compact(concat([try(aws_security_group.this[0].id, "")], var.vpc_security_group_ids))

   default_version                      = var.launch_template_default_version
   update_default_version               = var.update_launch_template_default_version
@@ -197,6 +197,7 @@ resource "aws_launch_template" "this" {
     for_each = var.network_interfaces
     content {
       associate_carrier_ip_address = lookup(network_interfaces.value, "associate_carrier_ip_address", null)
+      interface_type               = lookup(network_interfaces.value, "interface_type", null)
       associate_public_ip_address  = lookup(network_interfaces.value, "associate_public_ip_address", null)
       delete_on_termination        = lookup(network_interfaces.value, "delete_on_termination", null)
       description                  = lookup(network_interfaces.value, "description", null)
```
Although I ran into a new error. I was able to get past it by setting the variable to not create the Cloudwatch log group. I am not sure how the change is related to this new error.

`│ Error: Creating CloudWatch Log Group failed: ResourceAlreadyExistsException: The specified log group already exists:  The CloudWatch Log Group '/aws/eks/tf-west-test-vpc/cluster' already exists.`

## Versions
 - Terraform:
1.1.5
 - Provider(s):
AWS 3.74.2
 - Module:
 EKS 18.2.7

## Reproduction
Create an EKS cluster using the EKS module with a self_managed_node_group that has a network_interface defined. This will create a clean plan but when you apply the the autoscaling group will fail to create due to the issue listed above (When a network interface is defined for a launch template the security groups must be defined in the network interfaces section and not with vpc_security_group_ids in the main block of the launch template.)

### Code Snippet to Reproduce

```

self_managed_node_groups = { system = {
            name = "system"
            instance_type = var.node_type
            ami_id = var.ami.id
            desired_size = 2
            min_size = 1
            max_size = 4
            key_name = ""
            bootstrap_extra_args = "--kubelet-extra-args '--node-labels=node-role.jarvice.io/jarvice-system=true,node-pool.jarvice.io/jarvice-system=jxesystem --register-with-taints=node-role.jarvice.io/jarvice-system=true:NoSchedule'"
            network_interfaces = [
                    {
                      device_index = 0
                      associate_public_ip_address = true
                      interface_type = "efa"
                      security_groups = ["${aws_security_group.ssh.id}"] 
                    }
                ]
            enable_bootstrap_user_data = true
            pre_bootstrap_user_data = <<EOF
# pre_userdata (executed before kubelet bootstrap and cluster join)
# Add authorized ssh key
echo "${var.ssh_public_key}" >>/home/ec2-user/.ssh/authorized_keys
EOF
        }
}
```

## Expected behavior
Successful creation of a self managed node group with an EFA interface

## Actual behavior
Error on terraform apply.
│ Error: Error creating Auto Scaling Group: InvalidQueryParameter: Invalid launch template: When a network interface is provided, the security groups must be a part of it.

**Comment by bryantbiggs** [comment]
I don't think there is anything for us to do here, no? The error message from the API is pretty clear - a security group must be provided when creating a custom ENI - what solution are you looking for?

**Comment by mark-nimbix** [comment]
The EKS module is creating the bad launch template. I can not get around that using the EKS module.

**Comment by bryantbiggs** [comment]
> The EKS module is creating the bad launch template. I can not get around that using the EKS module.

I don't follow

**Comment by bryantbiggs** [comment]
from the upstream issue you linked - https://github.com/hashicorp/terraform-provider-aws/issues/4570#issuecomment-612201179 this is the intended behavior from the AWS API

**Comment by mark-nimbix** [comment]
> > The EKS module is creating the bad launch template. I can not get around that using the EKS module.
> 
> I don't follow

If you read the issue I linked you will see the error is misleading. To create a valid template you do need to define the security groups in the network_interfaces block which I have. But if you also set the vpc_security_groups_ids value in the launch_template you will get the error in the listed issue. So the problem is the EKS sub module is defining the vpc_security_groups_ids value always which creates an invalid template should you define any network interface.

**Comment by bryantbiggs** [comment]
ya, I don't see this as an issue with the module here. to get around this its quite simple, create your own launch template and pass it into the module

**Comment by mark-nimbix** [comment]
By that logic you might as well remove the network interfaces block from your help docs in the EKS module because per the issue we have been discussing you need to define a security group for a network interface and doing so will require that the vpc_security_groups_ids value be null in the launch_template. So from what I can tell this module can not use the network interfaces in any way with the self_managed_node_groups. 
The reason I did not create a launch template is because I am using bootstrap_extra_args which from what I can tell in the user_data module is dependent on the EKS cluster creation.

**Comment by bryantbiggs** [comment]
> By that logic you might as well remove the network interfaces block from your help docs in the EKS module because per the issue we have been discussing you need to define a security group for a network interface and doing so will require that the vpc_security_groups_ids value be null in the launch_template. So from what I can tell this module can not use the network interfaces in any way with the self_managed_node_groups. The reason I did not create a launch template is because I am using bootstrap_extra_args which from what I can tell in the user_data module is dependent on the EKS cluster creation.

correct - its dependent on *cluster creation, not node creation (needs endpoint, certificate, etc.). you can use the sub-modules separate from the root module though, this is how the root module is composed

**Comment by mark-nimbix** [comment]
> > By that logic you might as well remove the network interfaces block from your help docs in the EKS module because per the issue we have been discussing you need to define a security group for a network interface and doing so will require that the vpc_security_groups_ids value be null in the launch_template. So from what I can tell this module can not use the network interfaces in any way with the self_managed_node_groups. The reason I did not create a launch template is because I am using bootstrap_extra_args which from what I can tell in the user_data module is dependent on the EKS cluster creation.
> 
> correct - its dependent on *cluster creation, not node creation (needs endpoint, certificate, etc.). you can use the sub-modules separate from the root module though, this is how the root module is composed

So there is no way to use the network_interfaces in a self_managed_node_group with the current EKS module?

**Comment by mark-nimbix** [comment]
> I don't know the answer to that question - I haven't tried it myself. I do find it hard to believe since we have this module as well as the autoscaling module https://github.com/bryantbiggs/terraform-aws-autoscaling/blob/master/main.tf#L172-L191 and this is the first time I am hearing about this. I am sure there is a way to make it work, you'll just have to get creative and figure out how

Perhaps you are confused since I included EFA. If you remove the need for EFA the problem still exists. The network_interfaces block is unusable in the self_manage_node_groups.

**Comment by chuajiesheng** [comment]
It seems to be because `create_security_group` is defaulted to `true`.

Which resulted in [`self-managed-node-group.aws_security_group.this`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/modules/self-managed-node-group/main.tf#L441) being created.

And when [`self-managed-node-group.aws_launch_template.this`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/modules/self-managed-node-group/main.tf#L47) is run, it resulted in `vpc_security_group_ids` being populated with the `aws_security_group.this`.

And from [Launch template doc here](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#vpc_security_group_ids)

> [vpc_security_group_ids](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#vpc_security_group_ids) - A list of security group IDs to associate with. Conflicts with network_interfaces.security_groups

You can fix this by manually going to the UI and create a new Launch Template version, move the common security group out and update the ASG.

**Comment by mark-nimbix** [comment]
> It seems to be because `create_security_group` is defaulted to `true`.
> 
> Which resulted in [`self-managed-node-group.aws_security_group.this`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/modules/self-managed-node-group/main.tf#L441) being created.
> 
> And when [`self-managed-node-group.aws_launch_template.this`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/modules/self-managed-node-group/main.tf#L47) is run, it resulted in `vpc_security_group_ids` being populated with the `aws_security_group.this`.
> 
> And from [Launch template doc here](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#vpc_security_group_ids)
> 
> > [vpc_security_group_ids](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#vpc_security_group_ids) - A list of security group IDs to associate with. Conflicts with network_interfaces.security_groups
> 
> You can fix this by manually going to the UI and create a new Launch Template version, move the common security group out and update the ASG.

Thanks for the explanation. We ended up using a fork with a variation of the code fix I mentioned above. That ended up being a cleaner solution for us.

**Comment by antonbabenko** [comment]
This issue has been resolved in version 18.18.0 :tada:
